src/charm.py

```diff
@@ -13,91 +13,189 @@
 
 
 class MlflowCharm(CharmBase):
     """Charm the service."""
-
     _stored = StoredState()
 
     def __init__(self, *args):
         super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready)
+        # set interfaces
+        try:
+            self.interfaces = get_interfaces(self)
+        except NoVersionsListed as error:
+            self.model.unit.status = WaitingStatus(str(error))
+            return
+        except NoCompatibleVersions as error:
+            self.model.unit.status = BlockedStatus(str(error))
+            return
+        # install operator and prepare services
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.server_pebble_ready, self._on_server_pebble_ready)
+        # configurations
         self.framework.observe(self.on.config_changed, self._on_config_changed)
-        self.framework.observe(self.on.fortune_action, self._on_fortune_action)
-        self._stored.set_default(things=[])
-
-    def _on_httpbin_pebble_ready(self, event):
-        """Define and start a workload using the Pebble API.
-
-        TEMPLATE-TODO: change this example to suit your needs.
-        You'll need to specify the right entrypoint and environment
-        configuration for your specific workload. Tip: you can see the
-        standard entrypoint of an existing container using docker inspect
-
-        Learn more about Pebble layers at https://github.com/canonical/pebble
-        """
-        # Get a reference the container attribute on the PebbleReadyEvent
-        container = event.workload
-        # Define an initial Pebble layer configuration
-        pebble_layer = {
-            "summary": "httpbin layer",
-            "description": "pebble config layer for httpbin",
+        # actions
+        self.framework.observe(self.on.db_upgrade_action, self._dp_upgrade_action)
+        # relations
+        self.framework.observe(self.on.mysql_relation_changed, self._on_mysql_relation_changed)
+        self.framework.observe(self.on.mysql_relation_broken, self._on_mysql_relation_changed)
+        self.framework.observe(self.on.object_storage_relation_changed,
+                               self._object_storage_relation_changed)
+        self.framework.observe(self.on.object_storage_relation_departed,
+                               self._object_storage_relation_changed)
+        # initialise stored state
+        self._stored.set_default(
+            backend_store_uri=DEFAULT_BACKEND_STORE_URI,
+            artifact_root=DEFAULT_ARTIFACT_ROOT,
+            minio_environment={})
+        # initialise ingress
+        self.ingress = IngressRequires(self, {
+            "service-hostname": self.config["host"],
+            "service-name": self.app.name,
+            "service-port": self.config["port"]
+        })
+
+    def _on_install(self, _):
+        """Install on charm."""
+        self.unit.status = ActiveStatus()
```

Comment:
Should this be in `WaitingStatus`? It seems that pebble would need to be called to finish some configs, so the charm wouldn't be ready?